### DISCUSSION TITLE
GUI operations extremely slow with long video

### DISCUSSION THREAD

**Discussion by zetyang** [talmolab/sleap#757] [Help!] [answered]
Dear Sleap team,

I am using Sleap in a 3 hour video. When I try to check and fix the predictions in the GUI, I found all operations related to instances (e.g. change predicted instances to new instances by double clicking; delete instances; add new instances) become extremely slow. It takes about 1 min for each of these operations and the GUI would freeze meanwhile.

Do you know if this is normal or is there any way to fix this? Thank you so much!

**Comment by roomrys** [comment, marked as answer]
Hi @zetyang,

Do you know which version of SLEAP you are using? We recently released a new version of [SLEAP v1.2.3](https://github.com/talmolab/sleap/releases/tag/v1.2.3) which includes Labeling Suggestions Speed-Up (#709, #695).

Let us know if upgrading helps - or if you are already on the most current version of SLEAP, then this could be related to something else. Either way please let us know!

Thanks,
Liezl

**Reply by zetyang** [reply]
Hi @roomrys,

Thanks for your reply! I am using the newest version, v1.2.3.  It seems to be just related to operations with instances. I could scroll through the frames smoothly. But whenever I want to add/delete/modify an instance, it becomes extremely slow.

**Reply by roomrys** [reply]
Hi @zetyang,

Changing the instances in the project triggers a chain of callbacks - some of which require SLEAP to look-up project instances. In v1.2.3, we utilize a data cache to speed-up this look-up process; however, from your observation, it seems there is still some inefficiency proportional to the number of instances in the project - which we would like to fix/decouple!

Are you predicting on the entire video or on a subset of "Labeling Suggestions"? If you are using the "Labeling Suggestions", how many Labeling Suggestions do you have?

As a test, do you mind trying to predict on a smaller subset of "Labeling Suggestions" within your project (if you are not doing so already)?

Thanks,
Liezl

**Reply by zetyang** [reply]
Thanks a lot for your detailed reply, Liezl!

I am predicting on the entire 3hour video (25 fps). I also tried with shorter clips from the whole video. With 10min video, the GUI runs smoothly. But when the length goes up to 1hour, I start having this slow-down problem.

My main purpose here is to load all the predictions and manually correct the tracked nodes that were mislocated. I don't need to re-train the model afterwards. Is there a way that I could achieve this without any operations on the instances? 

One possible solution is to convert all the predicted instances in a slp file into labeled instances through command line. I tried to use the "sleap-convert" command by sending a predicted slp file as input and indicating another slp file as output. Unfortunately, it seems that the "sleap-convert" save instances as predicted instances into slp files by default. If there could an option in "sleap-convert" to allow saving instances as labeled ones, then my problem would be solved.

**Reply by roomrys** [reply]
Hi @zetyang,

For proofreading (correcting tracks), you do not need to convert the predictions to user instances; however, for "manually correcting nodes that were mislocated" you need user-labeled instances. I created a [notebook that will convert all predicted instances to user-labeled instances](https://colab.research.google.com/drive/1RcqtW6_AoPInbMuzKJN_g34YkHGv9ztW?usp=sharing) in your SLEAP Project file (.slp) for your purposes. However, we caution against using this notebook if you are retraining (as all user-labeled instances will be used for training - best to manually inspect each one carefully).

Let us know if this is what you were looking for.

We would still like to create a more permanent solution to the laggy behavior.

Thanks,
Liezl

The main code block from the aforementioned notebook:
```py
import sleap

from sleap.io.dataset import Labels
from sleap.gui.commands import AddUserInstancesFromPredictions


# Load the Labels object from the SLEAP Project file
labels: Labels = Labels.load_file(fn)
print("\n\nBefore creating user instances from predicted instances:")
print(f"\nlabels = {labels}")
print(f"len(labels.user_instances) = {len(labels.user_instances)}")
print(f"len(labels.predicted_instances) = {len(labels.predicted_instances)}")

# Convert all predicted instances to user-labeled instance.
# (Derived from sleap.commands.AddUserInstancesFromPredictions.do_action())
for lf in labels.labeled_frames:
  # Only add predicted instances on frames that are not user-labeled already
  if not lf.has_user_instances:
    new_instances = []
    unused_predictions = lf.unused_predictions
    for predicted_instance in unused_predictions:
        new_instances.append(
            AddUserInstancesFromPredictions.make_instance_from_predicted_instance(predicted_instance)
        )

    # Add the instances
    for new_instance in new_instances:
        labels.add_instance(lf, new_instance)

print("\n---------------------------------------------------------------------")
print("\n\nAfter creating user instances from predicted instances:")
print(f"\nlabels = {labels}")
print(f"len(labels.user_instances) = {len(labels.user_instances)}")
print(f"len(labels.predicted_instances) = {len(labels.predicted_instances)}")

# Save the Labels object to a SLEAP Project file
new_fn = "slp_all-user-instances"
labels.save(new_fn)

# Download the new SLEAP Project file
files.download(f"{new_fn}.slp")
```

**Reply by zetyang** [reply]
Thank you so much for your help, Liezl! This is extremely helpful. I actually started writing some code to replace all the predicted instances in a label object with normal instances, but experienced some difficulty saving the modified label object into a slp file. I guess the way I modified the labels just broke the object somehow. Your code works perfectly and solved this saving problem. Probably I should've used the "add_instance" function instead of just reassign the instances in each of the frame.

I will create a zip file including a test video and all the other related files that could replicate the GUI issue. Will send it along soon.

Thanks again for your very fast responses and for writing the code for converting the instances!
